### ISSUE TITLE
Apps in Multi-User are not showing up

### ISSUE THREAD

**Issue by TheMysticS** [d4rken-org/mystic-permissions#62]
I created a multi-user and installed a total of 8 apps.

Unfortunately, Mystic Permissions isn't showing them in the Apps Tab.

However, interestingly, when sorted by last updated criteria, it is showing these apps on top:

![IMG_20220720_204807](https://user-images.githubusercontent.com/54448474/180020602-f1bb7b25-da47-4e67-bbbc-d56d8bfdb62f.jpg)

Note that some of these apps are also installed in the Work Profile, e.g. File Manager+, DDG and SAI.

These 3 apps show the briefcase icon with number 1 inside them. It should be 2 since there are 2 clones (total of 3 installations including Main Profile).

The 2nd clone doesn't show up in Search either:

[Screenshot_2022-07-20-20-50-32-40_ce34de6323cfaeb7a2880d42987990b1](https://user-images.githubusercontent.com/54448474/180022137-e185c145-a38e-4545-8812-50a1125d73f8.jpg)

Likewise, apps installed in Multi-User but not in Work Profile don't show the briefcase badge.

I'll do another test. I'll install an app in multi-user that isn't installed in any other profile. I'll report back with whether Mystic Permissions is detecting that app from the main profile or not.

**Comment by TheMysticS** [comment]
Good news:

![IMG_20220720_211537](https://user-images.githubusercontent.com/54448474/180026032-894bc1cc-a2bb-49f1-b7a2-531b2ad8de7e.jpg)

I installed a new app called 'List My Apps in the multi-user. This isn't installed in any other profile.

The app shows up in the list right on top. This is the same problem we had earlier with apps installed in the Work Profile.

This is solvable. 😍😍

**Comment by d4rken** [comment]
There are 3 sources for app data currently:

https://github.com/d4rken-org/mystic-permissions/blob/43f2afb6e5e6bf2a56f0f436d45e22077a9acf37/app/src/main/java/eu/darken/myperm/apps/core/AppRepo.kt#L37-L43

* From the normal PackageManager without flags: Gets me all apps that are in the main profile
* From the normal PackageManager with `GET_UNINSTALLED`: Gets me all apps that are in the main profile and in other profiles
* From AppsLauncher: Gets me all? apps that are in the work profile

Processing for siblings and clones happens here:
https://github.com/d4rken-org/mystic-permissions/blob/43f2afb6e5e6bf2a56f0f436d45e22077a9acf37/app/src/main/java/eu/darken/myperm/apps/core/AppRepo.kt#L45-L80

If you only install an app in a secondary profile, and that profile is not the work profile, then I will get it from the package manager using `GET_UNINSTALLED`.  If i get it with and without the `GET_UNINSTALLED` flag, then this call here will assume it is the same app as the one in the main profile:

https://github.com/d4rken-org/mystic-permissions/blob/43f2afb6e5e6bf2a56f0f436d45e22077a9acf37/app/src/main/java/eu/darken/myperm/apps/core/AppRepo.kt#L64

My identification is always packagename+userhandle, but I currently only get a different userhandle for apps from AppsLauncher...

This is tricky... I'm not sure if this is possible. I think I can get a list of user handles, but there is no option to ask the system to filter by it:
https://developer.android.com/reference/android/content/pm/PackageManager

**Comment by TheMysticS** [comment]
The app installed exclusively in multi-users is showing up in the main profile. This is a very good thing.

So while getting it's icon may be difficult, it should be possible to mark them as apps in multi-user, at least for apps that are installed exclusively in the multi-user.

Only thing that remains to be seen is how to find apps that have clones in multi-users.

**Comment by d4rken** [comment]
> at least for apps that are installed exclusively in the multi-user.

Yes it should be the PackageInfo that we got when using `GET_UNINSTALLED` and not without, and it also is not in the data from AppLauncher.

**Comment by TheMysticS** [comment]
Alternative and simpler title:

**Recognize packages in Multi-Users.**

**Comment by TheMysticS** [comment]
Work Profile is just multi-user with a fancy name:

https://github.com/d4rken-org/mystic-permissions/issues/20#issuecomment-1186588930

While Work Profile and Multi-Users have some distinct differences, the way they are created is basically the same (new user with numbers 10, 11, 12, etc.).

Assuming user 10 is Work Profile, and apps in that user can be listed, how different are the other users (multi-users) from an implementation point of view?

**Comment by d4rken** [comment]
> Assuming user 10 is Work Profile, and apps in that user can be listed, how different are the other users (multi-users) from an implementation point of view?

That depends on how we get the information.

Currently the workprofile info comes from `LauncherApps` service, which can returns a list of profiles, and we are iterating that (minus the currently process userhandle which i assume would be UserHandle-0):

https://github.com/d4rken-org/mystic-permissions/blob/43f2afb6e5e6bf2a56f0f436d45e22077a9acf37/app/src/main/java/eu/darken/myperm/apps/core/container/SecondaryProfilePkg.kt#L69-L108

**Comment by TheMysticS** [comment]
So would it be right to say that Work Profile is a special type of Multi-user where User 0 (Default user), including apps installed in the main user, has some additional permissions and access, without root?

**Comment by d4rken** [comment]
FYI despite the second user, `LauncherApps` just returns the work profile:

```java
launcherApps = {LauncherApps@24622} 
profiles = {ArrayList@24623}  size = 2
 0 = {UserHandle@24639} "UserHandle{0}"
 1 = {UserHandle@24636} "UserHandle{10}"
```

Accessing e.g. `UserManager.userCount` requires yields a SecurityException: `You either need MANAGE_USERS or CREATE_USERS permission to: query users`

`UserManager.userProfiles` works without permission but again only returns:


```java
result = {ArrayList@24730}  size = 2
 0 = {UserHandle@24639} "UserHandle{0}"
 1 = {UserHandle@24636} "UserHandle{10}"
```

the second user should be `11`:

```
sargo:/data/user # ls
0  10  11
```

**Comment by d4rken** [comment]
I manually created a `UserHandle` for `11` and when trying to access the apps via `LauncherApps` service, I get a `SecurityException`: 
`Cannot retrieve activities for unrelated profile 11`

**Comment by d4rken** [comment]
I think this is working as good as we can get it without going for root, there might be a way to get more info without root, but using a shell, but this is a huge change that is low priority.
